AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStoreAdmin/src/CreateKeys.dfy

```diff
@@ -0,0 +1,400 @@
+                                                                           AwsCryptographyPrimitives := e
+                                                                         ));
+
+    // Pack BKC Digest & Plaintext
+    var activePlaintextTuple := HvUtils.PackPlainTextTuple(activeDigest, activePlaintextMaterial);
+    var decryptOnlyPlaintextTuple := HvUtils.PackPlainTextTuple(decryptOnlyDigest, activePlaintextMaterial);
+    var beaconPlaintextTuple := HvUtils.PackPlainTextTuple(beaconDigest, beaconPlaintextMaterial);
+
+    assert KMSKeystoreOperations.AttemptKmsOperation?(kmsConfiguration, decryptOnlyBranchKeyContext[Structure.KMS_FIELD]);
+    var wrappedDecryptOnlyBranchKey? := KMSKeystoreOperations.EncryptKey(
+      decryptOnlyPlaintextTuple,
+      customEncryptionContext,
+      decryptOnlyBranchKeyContext[Structure.KMS_FIELD],
+      kmsConfiguration,
+      grantTokens,
+      kmsClient
+    );
+    var wrappedDecryptOnlyBranchKey :- wrappedDecryptOnlyBranchKey?.MapFailure(e => ConvertKmsErrorToError(e));
+    assert KMSKeystoreOperations.AttemptKmsOperation?(kmsConfiguration, activeBranchKeyContext[Structure.KMS_FIELD]);
+    var wrappedActiveBranchKey? := KMSKeystoreOperations.EncryptKey(
+      activePlaintextTuple,
+      customEncryptionContext,
+      activeBranchKeyContext[Structure.KMS_FIELD],
+      kmsConfiguration,
+      grantTokens,
+      kmsClient
+    );
+    var wrappedActiveBranchKey :- wrappedActiveBranchKey?.MapFailure(e => ConvertKmsErrorToError(e));
+    assert KMSKeystoreOperations.AttemptKmsOperation?(kmsConfiguration, beaconBranchKeyContext[Structure.KMS_FIELD]);
+    var wrappedBeaconKey? := KMSKeystoreOperations.EncryptKey(
+      beaconPlaintextTuple,
+      customEncryptionContext,
+      beaconBranchKeyContext[Structure.KMS_FIELD],
+      kmsConfiguration,
+      grantTokens,
+      kmsClient
+    );
+    var wrappedBeaconKey :- wrappedBeaconKey?.MapFailure(e => ConvertKmsErrorToError(e));
+    // Write ACTIVE, Version & Beacon Items to storage
+    var writeItems? := storage.WriteNewEncryptedBranchKey(
+      KeyStoreTypes.WriteNewEncryptedBranchKeyInput(
+        Active := Structure.ConstructEncryptedHierarchicalKey(
+          activeBranchKeyContext,
+          wrappedActiveBranchKey
+        ),
+        Version := Structure.ConstructEncryptedHierarchicalKey(
+          decryptOnlyBranchKeyContext,
+          wrappedDecryptOnlyBranchKey
+        ),
+        Beacon := Structure.ConstructEncryptedHierarchicalKey(
+          beaconBranchKeyContext,
+          wrappedBeaconKey
+        )
+      )
+    );
+    var _ :- writeItems?
+    .MapFailure(e => Types.AwsCryptographyKeyStore(
+                    AwsCryptographyKeyStore:= e
+                  ));
+
+    output := Success(
+      Types.CreateKeyOutput(
+        Identifier := branchKeyIdentifier,
+        HierarchyVersion := hierachyVersion
+      ));
+  }
+
+  twostate predicate WrappedBranchKeyCreationHV2?(
```

Comment:
Note for @texastony : I need to come back to this and review what a twostate predicate is; my Dafny knowledge is not deep enough.